// common/ota_service.proto
syntax = "proto3";

package ota;

// Service principal OTA
service OTAService {
    // Vérifier les mises à jour disponibles
    rpc CheckUpdate(UpdateRequest) returns (UpdateResponse);
    
    // Télécharger un fichier de mise à jour
    rpc DownloadUpdate(DownloadRequest) returns (stream FileChunk);
    
    // Confirmer l'installation
    rpc ConfirmInstallation(InstallationRequest) returns (InstallationResponse);
    
    // Obtenir le statut d'un dispositif
    rpc GetDeviceStatus(DeviceStatusRequest) returns (DeviceStatusResponse);
    
    // Enregistrer un dispositif
    rpc RegisterDevice(DeviceRegistration) returns (RegistrationResponse);
}

// Messages de requête
message UpdateRequest {
    int32 device_id = 1;
    string current_version = 2;
    UpdateType update_type = 3;
}

message DownloadRequest {
    int32 device_id = 1;
    int32 update_id = 2;
    int64 chunk_offset = 3;
}

message InstallationRequest {
    int32 device_id = 1;
    int32 update_id = 2;
    UpdateStatus success = 3;
    DeviceStatus status = 4;
    string error_message = 5;
}

message DeviceStatusRequest {
    int32 device_id = 1;
}

message DeviceRegistration {
    int32 device_id = 1;
    string device_type = 2;
    string current_version = 3;
    string platform = 4;
}

// Messages de réponse
message UpdateResponse {
    bool update_available = 1;
    int32 update_id = 2;
    string version = 3;
    string description = 4;
    int64 file_size = 5;
    string checksum = 6;
    UpdateType update_type = 7;
}

message FileChunk {
    bytes data = 1;
    int64 offset = 2;
    bool is_last = 3;
}

message InstallationResponse {
    UpdateStatus success = 1;
    string message = 2;
}

message DeviceStatusResponse {
    int32 device_id = 1;
    string current_version = 2;
    string last_update = 3;
    DeviceStatus status = 4;
}

message RegistrationResponse {
    bool success = 1;
    string message = 2;
}

// Énumérations
enum UpdateType {
    CONFIG_FILE = 0;
    APPLICATION = 1;
    SYSTEMD_SERVICE = 2;
}

enum DeviceStatus {
    ONLINE = 0;
    OFFLINE = 1;
    UPDATING = 2;
    ERROR = 3;
}

enum  UpdateStatus {
    PENDING = 0;
    DOWNLOADING = 1 ;
    DOWNLOADED = 2 ;
    INSTALLING = 3;
    SUCCESS = 4;
    FAILED = 5;
    ROLLBACK = 6;
}
